### PR TITLE
feat(subpartition): finish sub partitions

### PR DESCRIPTION
**Pull request by zijiacj**
#### What type of PR is this?
type-feature

#### What this PR does / why we need it:
finish sub partitions and pass tests

#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/plugins/schema-plugin-ob-mysql/src/test/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParserTest.java [hunk 1]
```diff
@@ -230,7 +230,7 @@ public void getPartition_secondary_key_no_template_Success() {
         Assert.assertEquals("col3", partition.getSubpartition().getPartitionOption().getColumnNames().get(1));
         Assert.assertEquals(DBTablePartitionType.KEY, partition.getSubpartition().getPartitionOption().getType());
         Assert.assertEquals(2, partition.getSubpartition().getPartitionOption().getColumnNames().size());
-        Assert.assertTrue(partition.getSubpartition().getPartitionOption().getPartitionsNum() == 3);
+        Assert.assertNull(partition.getSubpartition().getPartitionOption().getPartitionsNum());
```
Comment (MarkPotato777) [inline server/plugins/schema-plugin-ob-mysql/src/test/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParserTest.java:233]: What's this change for?

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 2]
```diff
@@ -244,33 +246,102 @@ public DBTablePartition getPartition() {
...
+                    subPartitionDefinitions.add(partitionDefinition);
+                }
+            } else {
+                // obtain DBTablePartitionDefinitions for template subpartitions
+                String parentPartitionName = removeIdentifiers(partitionElement.getRelation());
+                List<SubPartitionElement> templates = partitionStmt.getSubPartitionOption().getTemplates();
+                for (int i = 0; i < templates.size(); i++) {
+                    DBTablePartitionDefinition partitionDefinition = new DBTablePartitionDefinition();
+                    partitionDefinition.setName(
+                            parentPartitionName + 's' + removeIdentifiers(templates.get(i).getRelation()));
```
Comment (MarkPotato777) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:338]: what does `+ 's'` mean?

File: server/plugins/schema-plugin-ob-mysql/src/test/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParserTest.java (hunk 1 above)
Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/test/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParserTest.java:233]: Non-template subpartition should not have partitionNum
image

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 3]
```diff
@@ -244,33 +246,102 @@ public DBTablePartition getPartition() {
...
+            parseSubPartition(subPartitionOption, subPartitionDefinitions, partitionStmt,
+                    DBTablePartitionType.RANGE_COLUMNS, false);
+        } else if ("list".equals(type.toLowerCase())) {
+            parseSubPartition(subPartitionOption, subPartitionDefinitions, partitionStmt, DBTablePartitionType.LIST,
+                    true);
+        } else if ("list columns".equals(type.toLowerCase())) {
+            parseSubPartition(subPartitionOption, subPartitionDefinitions, partitionStmt,
+                    DBTablePartitionType.LIST_COLUMNS, false);
+        } else {
+            partition.setWarning("Only support HASH/KEY subpartition currently");
```
Comment (MarkPotato777) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:270]: The log should be updated.

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 4]
```diff
@@ -244,33 +246,102 @@ public DBTablePartition getPartition() {
...
+        }
+        return partition;
+    }
+
+    private void parseSubPartition(DBTablePartitionOption subPartitionOption,
```
Comment (MarkPotato777) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:275]: Actually, this method is not related to `parse` itself. `FillSubPartition` may be more clear.

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 5]
```diff
@@ -244,33 +246,102 @@ public DBTablePartition getPartition() {
...
-                    .setPartitionsNum(partition.getSubpartitionTemplated() ? subOption.getTemplates().size()
-                            : partitionStmt.getPartitionElements().get(0).getSubPartitionElements().size());
+            parseSubPartition(subPartitionOption, subPartitionDefinitions, partitionStmt, DBTablePartitionType.KEY,
+                    false);
         } else if ("hash".equals(type.toLowerCase())) {
-            subPartitionOption.setType(DBTablePartitionType.HASH);
-            Expression expression = subOption.getSubPartitionTargets().get(0);
+            parseSubPartition(subPartitionOption, subPartitionDefinitions, partitionStmt, DBTablePartitionType.HASH,
+                    true);
+        } else if ("range".equals(type.toLowerCase())) {
```
Comment (MarkPotato777) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:257]: Could we eliminate these `if` branches? Actually, your `parseSubPartition` method only cares about the `DBTablePartitionType` and if it supports `expression`. 
1. May be you could convert the String to Enum in the Enum itself, like do it in  `DBTablePartitionType#fromValue`.
2. The answer to whether the partition type supports expression could also be implied in the `DBTablePartitionType`.
In this way, you could eliminate all the if braches and directly call `parseSubPartition`.

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java (hunk 3 above)
Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:270]: ok

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java (hunk 4 above)
Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:275]: ok

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java (hunk 5 above)
Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:257]: ok

**Review by PeachThinking — APPROVED** [review]
lgtm

**Review by MarkPotato777 — APPROVED** [review]
LGTM

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 6]
```diff
@@ -244,33 +248,90 @@ public DBTablePartition getPartition() {
...
-            subPartitionOption.setColumnNames(subOption.getSubPartitionTargets() == null ? null
-                    : subOption.getSubPartitionTargets().stream().map(item -> removeIdentifiers(item.getText()))
-                            .collect(Collectors.toList()));
-            subPartitionOption
-                    .setPartitionsNum(partition.getSubpartitionTemplated() ? subOption.getTemplates().size()
-                            : partitionStmt.getPartitionElements().get(0).getSubPartitionElements().size());
-        } else if ("hash".equals(type.toLowerCase())) {
-            subPartitionOption.setType(DBTablePartitionType.HASH);
-            Expression expression = subOption.getSubPartitionTargets().get(0);
+        DBTablePartitionType dbTablePartitionType = DBTablePartitionType.fromValue(type.toUpperCase());
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:261]: what about range and range column, list and list columns?

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 7]
```diff
@@ -244,33 +248,90 @@ public DBTablePartition getPartition() {
...
+                    DBTablePartitionDefinition partitionDefinition = new DBTablePartitionDefinition();
+                    partitionDefinition.setName(
+                            removeIdentifiers(partitionElement.getSubPartitionElements().get(i).getRelation()));
+                    partitionDefinition.setOrdinalPosition(i);
+                    partitionDefinition.setType(dbTablePartitionType);
+                    subPartitionDefinitions.add(partitionDefinition);
+                }
+            } else {
+                // obtain DBTablePartitionDefinitions for templated subpartitions
+                String parentPartitionName = removeIdentifiers(partitionElement.getRelation());
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:321]: why should we set sub-partition in every partition def?

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 8]
```diff
@@ -244,33 +248,90 @@ public DBTablePartition getPartition() {
...
-        // TODO 目前 ODC 仅支持 HASH/KEY 二级分区, 其它类型后续需补充
+        fillSubPartitions(partition, partitionStmt);
+        return partition;
+    }
+
+    private void fillSubPartitions(DBTablePartition partition, Partition partitionStmt) {
         partition.setSubpartitionTemplated(partitionStmt.getSubPartitionOption().getTemplates() != null);
-        SubPartitionOption subOption = partitionStmt.getSubPartitionOption();
+        DBTablePartitionOption subPartitionOption = partition.getSubpartition().getPartitionOption();
+        List<DBTablePartitionDefinition> subPartitionDefinitions =
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:258]: will it be null？

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 9]
```diff
@@ -244,33 +248,90 @@ public DBTablePartition getPartition() {
         if (partitionStmt.getSubPartitionOption() == null) {
             return partition;
         }
-        // TODO 目前 ODC 仅支持 HASH/KEY 二级分区, 其它类型后续需补充
+        fillSubPartitions(partition, partitionStmt);
+        return partition;
+    }
+
+    private void fillSubPartitions(DBTablePartition partition, Partition partitionStmt) {
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:255]: what about ob-oracle？

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java (hunk 8 above)
Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java]: image
not null

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java]: remove toUppercase()

File: server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java [hunk 10]
```diff
@@ -241,36 +236,140 @@ public DBTablePartition getPartition() {
...
+                    subPartitionDefinitions.add(subPartitionDefinition);
+                }
+            }
+        }
+    }
+
+    private void fillSubPartitionValue(DBTablePartitionType subDBTablePartitionType,
+            SubPartitionElement subPartitionElement,
+            DBTablePartitionDefinition subPartitionDefinition) {
+        if (subDBTablePartitionType == DBTablePartitionType.LIST) {
```
Comment (yhilmare) [inline server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:339]: don't use DBTablePartitionType, use `instance of` instead, it is more reliable:

```Java
if (subPartitionElement instanceof SubListPartitionElement) {
    ....
}
```

Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:339]: The object of the SubListPartitionElement still has to judge the processing logic against the subDBTablePartitionType. Because SubListPartitionElement contains both LIST and LIST_COLUMNS

Comment (zijiacj) [reply server/plugins/schema-plugin-ob-mysql/src/main/java/com/oceanbase/odc/plugin/schema/obmysql/parser/OBMySQLGetDBTableByParser.java:339]: ok

**Review by yhilmare — APPROVED** [review]
lgtm

**Review by PeachThinking — APPROVED** [review]

**Review by MarkPotato777 — APPROVED** [review]
